### ISSUE TITLE
self-healing: healthchecks within kairos-agent

### ISSUE THREAD

**Issue by mudler** [kairos-io/kairos#162]
**Is your feature request related to a problem? Please describe.**
Beside kubernetes reporting, it would be ideal to re-use the kairos agent to run periodically user-defined healthchecks.

**Describe the solution you'd like**
Healthchecks ideally should trigger execution of user-defined scripts (e.g. by "pseudo"-containers) to re-act and self-heal a node.

**Describe alternatives you've considered**
Run healthchecks within Kubernetes - but in case of Kubernetes failures there is nothing that can be done

**Additional context**
What I'm worried here is to provide still transparency on what's happening - while Kubernetes makes it transparent via API calls and logging accessible within it, here we need to make sure action takens can be tracked by humans.

**Comment by ci-robbot** [comment]
Triage (no comments here in four years, so starting from what the code says today).

## Nothing of this exists

`grep -rni 'healthcheck|greenboot|self-heal'` over `master` returns two hits, both in `provider/internal/services/edgevpn.go:38,53`, and both are edgevpn's own `--enable-healthchecks` API flag. There is no health daemon, no health policy type, no periodic evaluation, nothing that polls Kubernetes for policies. The issue is unimplemented as written.

## But one half of it arrived by another route, and only on one boot path

The greenboot primitive this issue is really after, "a bad boot undoes itself", exists for trusted boot:

- `utils.AddBootAssessment` (`agent/pkg/utils/common.go:639-674`) renames boot entries with a `+3` tries counter, which is systemd-boot's automatic assessment and rollback.
- It is called from exactly three places, all UKI: `agent/pkg/uki/install.go:188`, `agent/pkg/uki/upgrade.go:114`, `agent/pkg/uki/reset.go:98`.

On the grub path there is no equivalent. The code says so out loud, at `agent/internal/agent/hooks/gruboptions.go:49-51`:

```go
// TODO: There are more things written in the OEM partition to make boot assessment
// and next-boot selection work. These are not handled at all right now. We should
// fix those before we officially support OEM encryption.
```

So a non-UKI node that takes a bad upgrade does not roll back on its own.

## The remote half also arrived, as commands rather than policies

`agent/internal/phonehome` plus the AuroraBoot fleet server now gives a node a heartbeat and a command poll. The command allowlist is `agent/internal/phonehome/config.go:68`:

```go
var DefaultAllowedCommands = []string{"upgrade", "upgrade-recovery", "reboot", "unregister"}
```

That is remediation an operator can trigger from outside Kubernetes, which is the property this issue wanted ("in case of Kubernetes failures there is nothing that can be done"). What is missing is the automatic part: nothing on the node decides it is unhealthy, and the fleet server expresses commands, not health policies.

## Suggested rescope

As one epic this is not actionable. I would split it into three, in this order:

1. **grub-path boot assessment**, to match the UKI behaviour. This is the highest-value piece, the smallest, and it is the one users actually feel. It also unblocks the OEM-encryption TODO quoted above.
2. **local health policies in the agent**: user-defined checks, an interval, and a remediation action, evaluated on the node. This is the greenboot-shaped part and the only genuinely new design work.
3. **remote health policies**: express a policy over the existing phone-home channel instead of a one-shot command. Depends on 2 and is cheap once 2 exists.

Also worth noting: the `area/CAPI-shim` label predates the CAPI providers actually shipping (v0.1.0, see #120), and node health is not in their scope, so that label is misleading now.

@mudler you filed this, and the greenboot reference is still the right model. Want me to open 1 as its own issue and close this as the umbrella, or keep it as is?
